### PR TITLE
[pdfjs] WIP

### LINKED ISSUE #139: Add PDF.js
[PDF.js](http://mozilla.github.io/pdf.js/) is a full-blown PDF parser and viewer made completely in JavaScript and HTML5 by Mozilla. It's the very same PDF viewer which is currently integrated into Firefox.

Adding this into cljsjs would allow ClojureScript community to build on top of it more advanced PDF viewers for example Reagent, Om and so forth without them having to embed PDF.js into their source code.

### PR DESCRIPTION
**Pull request by Deraen**
Fixes #139 

@empperi Could you help testing this?

Notes and possible problems:
- What is the role of pdf.worker.js? Is it loaded by pdf.js?
- Are the files in `web/` directory needed?
- Looks like pdf.js doesn't distribute minifed files and it's hard to minify (might be possible with Closure, but asset-minified doesn't support setting the proper options)

### CODE REVIEW

**Comment by Deraen** [comment]
Ping @empperi.

If no one helps to test this and preferably provides fixes I don't see this moving ahead. Maybe one day I need this myself, but until then I don't have time to finish this.

**Comment by empperi** [comment]
Sorry about being lazy about this. Truth to be told, you made few questions I couldn't answer back then. Then privatizations moved to other things and haven't had time to look into this.

However, I've been in the past few days been doing PDF.js integration into our application and while doing that I've learned few things about the library. This will also help me to work on testing this pull request. Let me elaborate how PDF.js works.

PDF.js is put into three different layers, the first one is responsible of parsing the PDF files and turning them to executable JavaScript code, it's a bit like a PDF JIT compiler for JavaScript. The second part is the rendering engine, it performs the drawing into HTML5 canvas. The third part is the viewer integration which means stuff like zooming, overlaying text over the canvas so you can select text and so forth.

The first two are combined in [pdf.combined.js](https://github.com/mozilla/pdfjs-dist/blob/master/build/pdf.combined.js), the last part is in [pdf_viewer.js](https://github.com/mozilla/pdfjs-dist/blob/master/web/pdf_viewer.js). You can use PDF.js without the viewer part but then you are limited to just rendering PDFs into canvas - you need to take care of everything else by yourself, including providing text selection support which is no small feat.

So basically one could divide this into two parts, the one including the pdf.combined.js and the other with pdf_viewer.js (which of course has a dependency to pdf.combined.js).

I'll try to dig into this next week and hopefully find time to do some work on the externs definitions.

**Comment by viesti** [comment]
Realising that there is more to it, but the app we're working with was happy after adding the following to pdf.ext.js

```
0% diff -w pdf.ext.js xxx/src/externs/pdf.ext.js
149a150,153
>   "getPage": function () {},
>   "getViewport": function () {},
>   "then": function () {},
>   "cancel": function () {},
155c159,165
<     "SVGGraphics": function () {}
---
>   "SVGGraphics": function () {},
>   "PDFPageView": {
>     "setPdfPage": function () {},
>     "update": function () {},
>     "draw": function () {},
>   },
>   "DefaultTextLayerFactory": {}
```

**Comment by Deraen** [comment]
@viesti Could you paste the diff in `-u` format?

**Comment by viesti** [comment]
Didn't try it this patches but here's another try :)

```
0% diff -w -u pdf.ext.js mediaseuranta/src/externs/pdf.ext.js
--- pdf.ext.js	2015-09-14 14:46:14.000000000 +0300
+++ mediaseuranta/src/externs/pdf.ext.js	2015-09-14 14:39:47.000000000 +0300
@@ -147,10 +147,20 @@
     "maxCanvasPixels": {},
     "openExternalLinksInNewWindow": {},
     "getDocument": function () {},
+  "getPage": function () {},
+  "getViewport": function () {},
+  "then": function () {},
+  "cancel": function () {},
     "PDFDataRangeTransport": function () {},
     "Metadata": function () {},
     "AnnotationUtils": {
         "getHtmlElement": function () {}
     },
-    "SVGGraphics": function () {}
+  "SVGGraphics": function () {},
+  "PDFPageView": {
+    "setPdfPage": function () {},
+    "update": function () {},
+    "draw": function () {},
+  },
+  "DefaultTextLayerFactory": {}
 };
```

**Comment by Deraen** [comment]
Anyone using this? Would like to merge this but I don't have time to test this.

**Comment by empperi** [comment]
We do use it in our application. It's obviously a partial externs file but works at least in our context. Could double check at work before you merge it. Away from work for friday but remind me on monday and will check, promise :)

**Comment by Deraen** [comment]
@empperi ping :D

**Comment by empperi** [comment]
Right :)

Ok, checked this to the externs file we have and it works at least on our context. I'm 100% certain this isn't a complete externs file since the API for PDF.js is quite large but it allows at least us to receive PDF files, render them on canvas, apply a text layer on top of it and do a bunch of other stuff too. So it should be fine for most people.

So yes, this should work - at least partially.
